### ISSUE TITLE
Broadlink RM4 PRO is not able to learn or to send RF Signal

### ISSUE THREAD

**Issue by Virtual-Nomad** [mjg59/python-broadlink#358]
Unsure if this is the right place to raise this issue, however, I cam across https://github.com/mjg59/python-broadlink/issues/355 which relates to a similar issue except with IR. Currently, I am trying to get my RM4 Pro to work with the HA and having issues bring able to learn RF (IR learns and sends no problem at all) with both python-broadlink and Home assistant. (I also have an issue with it as a sensor (temperature and humidity) - although not that concerned/interested in this side at the moment)

Current Setup
Home Assistant 0.109.3
Broadlink RM4 Pro
Added to network using python-broadlink - PASS
when I attempt 
devices[0].host[0] > ‘192.168.1.172’ - PASS
devices[0].get_type() > ‘Unknown’ - Cant get the type (so this may be a concern or compatibility issue)
devices[0].enter_learning() > AttributeError: device instance has no attribute ‘enter_learning’ _FAIL

Under Home assistant 
Added to HA as switch - configuration.yaml
switch:
 platform: broadlink
 host: 192.168.1.172
 mac: 24:df:a7:b9:c0:5a
 type: rm4_pro
 timeout: 15
 retry: 5
 switches:
   tv_samsung_power:
   friendly_name: "Samsung TV Power"
     command_on: IRSTRING
     command_off: IRSTRING
When I run broadlink.learn
- Red LED comes on
- I can capture IR without an issue as I did with the Samsung TV Power
- **UNABLE TO CAPTURE ANY RF**

As far as the sensor is concerned
sensor:
 platform: broadlink 
 host: 192.168.1.172
 mac: 24:df:a7:b9:c0:5a
 type: rm4_pro
 monitored_conditions:
    temperature
    humidity
Both Temperature and Humidity just show a value of 0

Do you know if anyone has been successful with the RM4 Pro and RF?

Another point is that I cant get the RM4 Pro to work with e-control (or broadlink manager) so have not been able to get any RF packets to test it broadlink.send command to see if that works.

**Comment by felipediel** [comment]
@Virtual-Nomad The RF functionality has not been fully tested because I don't have the device. Would you like to help me by making some captures in Wireshark so I can add this functionality?

**Comment by Virtual-Nomad** [comment]
@felipediel more than happy to. Just be aware I am from the windows world and only now trying to get up to speed with all things python, home assistant. But I am familiar with wireshark (in windows world) so with a little guidance would be able to help out.

**Comment by felipediel** [comment]
@Virtual-Nomad Do you have an iOS device?

**Comment by Virtual-Nomad** [comment]
Yes mate (although I can arrange an android if required). Is there an alternate communications channel you wish to use for all testing results/process etc.  I am in Brisbane, Australia so time zone wise it is UTC +10 with evenings and weekends being the times I can devote to any testing.

**Comment by felipediel** [comment]
@Virtual-Nomad Great!
1. You need to run [this program](https://github.com/gh2o/rvi_capture) on Windows and make some captures while controlling the device with the official app on an iOS device. Execute only one type of command per capture and name the file with the name of the command (e.g. learn_rf).
2. Run [this script](https://drive.google.com/open?id=1Gr-Qe_sdVL-Yn2g2ITDFYB1TTpd1pX8Z) using `python3 debug.py`. A file named debug.txt will be created.
3. Send me the files.

**Comment by Virtual-Nomad** [comment]
So currently setup 
Windows 10
Installed Wireshark (Windows)
installed itunes
Installed python3
Downloaded  rvi_capture.py
Downloaded debug.py
- Ensured applemobiledeviceService.exe was running

First run of "rvi_capture.py" 
- On first execute it stated downloading libimobiledevice
Second test run rvi_capture.py - | "c:\Program Files\Wireshark\Wireshark.exe" -k -i -
- Opens wireshark in a capture mode so looking good (just need to get home to actually test)

when I run debug.py however I am getting this error ... do I need to download additional components?
C:\Temp\rvi_capture-master>python3 debug.py
Traceback (most recent call last):
  File "debug.py", line 40, in <module>
    from cryptography.hazmat.backends import default_backend
ModuleNotFoundError: No module named 'cryptography'

**Comment by Virtual-Nomad** [comment]
![image](https://user-images.githubusercontent.com/57859302/81157316-309d4380-8fca-11ea-9d48-2611bb5bd502.png)
So as per the above (with the exception of debug) I think I have everything running. 
Connected the iphone to the PC via the USB
Installed the Broadlink App
- Added the RM4_Pro into via AP mode
- Added an appliance using the UserDefined, set the room to Office, Named the Device: Office Fan
- Added a function button > Start Learning 
Laptop cmd - Ran the command "rvi_capture.py - | "c:\Program Files\Wireshark\Wireshark.exe" -k -i -" to start capturing packets
Broadlink App
- Started learning (I have a physical remote)  - LED on RM4 Showed Up 
- Long Pressed the Low Fan button on the RF remote - LED turned off (fan turned on low)
- Broadlink App prompted for send short press - LED on RM4 Showed Up 
- Short Press the Low Fan - LED turned off
NB: As per the screen shot wireshark shows No Packets captured 
Laptop - Exited wireshark
Laptop command - python3 debug.py
- Traceback (most recent call last):
-   File "debug.py", line 40, in <module>
-     from cryptography.hazmat.backends import default_backend
- ModuleNotFoundError: No module named 'cryptography'

No files created (although there is an empty  __pycache__ sub directory) 

I assume I am missing something basic (NB I have also removed webroot incase it was an issue)

**Comment by Virtual-Nomad** [comment]
@felipediel This may not actually achieve what it is you are after, however, given I wasn't really sure how the requested approach worked. I setup a hotspot AP on my laptop and connected both the iphone and RM4 Pro to the Wifi.
I then used wireshark to record all traffic across the hotspot wifi during the process of learning an RF signal. It will probably have a lot of there traffic captured but it may hold what you are after. in this capture the iphone is 192.168.137.152 and the RM4 is  192.168.137.126
[RFOfficeFan-LearnLowSpeed.zip](https://github.com/mjg59/python-broadlink/files/4586201/RFOfficeFan-LearnLowSpeed.zip)

**Comment by Virtual-Nomad** [comment]
Side note ... I just noticed rvi_capture states "NOTE: There has been a major protocol change in iOS 13, so this will only work for iOS 12 and earlier." my iphone is IOS 13. Not sure if that has factored.

**Comment by felipediel** [comment]
@Virtual-Nomad Pure Wireshark does not solve. If you did not succeed with rvictl you can install the Broadlink application on an Android emulator and make the captures with Wireshark. I recommend [MEmu emulator](https://www.memuplay.com/). You may find [this comment](https://github.com/home-assistant/core/issues/30215#issuecomment-593896024) useful.

To solve the cryptography error just type:
`pip install cryptography`

**Comment by Virtual-Nomad** [comment]
@felipediel wasn't able to install MEmu as just get a run time error. Anyway found an older iphone with IOS12 so managed to revert back to RVICTL. so I have attached both learning and sending captures. 
Not sure about the debug.txt but I have included it
I am correct you simply run python debup.py once you have finished the capture? 
[RFOfficeFan-LearningOff.zip](https://github.com/mjg59/python-broadlink/files/4592188/RFOfficeFan-LearningOff.zip)
[RFOfficeFan-SendingOff.zip](https://github.com/mjg59/python-broadlink/files/4592189/RFOfficeFan-SendingOff.zip)

**Comment by felipediel** [comment]
@Virtual-Nomad Great! The captures are good. Now I need you to [disable the cloud](https://github.com/home-assistant/core/issues/30215#issuecomment-581159499), run debug.py again and send me debug.txt.

**Comment by Virtual-Nomad** [comment]
Not sure if I am clear what I need to do however;
- I have removed the device from the app (device went into discover mode)
- Ran Debug.py - 
[debug-PostRemovalfromBroadlinkApp.txt](https://github.com/mjg59/python-broadlink/files/4596241/debug-PostRemovalfromBroadlinkApp.txt)

Re-established direct connection/setup using python-broadlink 
>>> import broadlink
>>> broadlink.setup('','',3)
>>> devices = broadlink.discover(timeout=5)
>>> devices[0].auth()
True
>>> devices[0].host
('192.168.137.220', 80)
- Ran Debug.py - 
[debug-post-Python-BoradlinkImport.txt](https://github.com/mjg59/python-broadlink/files/4596256/debug-post-Python-BoradlinkImport.txt)

I did go to the disable the cloud link above but wasn't really sure what was required. So hoping I read between the lines correctly

**Comment by felipediel** [comment]
Sorry, I was unable to decrypt the content of the messages because your device was connected to the cloud during the captures. You need to block the device from accessing the internet using a firewall. Here are the [instructions](https://github.com/home-assistant/core/issues/30215#issuecomment-625566467) from someone who has achieved this before.

**Comment by Virtual-Nomad** [comment]
@felipediel just so I am clear should I be using the python-broadlink integration to simply establish a connection to the device and run the debug.py. I guess follow what i did above but simply be disconnected from the internet during this the tests?

**Comment by felipediel** [comment]
@Virtual-Nomad The debug serves to intercept your local key. It worked, you don't need to run it anymore.

The problem is with the captures. You need to block the IP address of the Broadlink device using a firewall (your router probably has a built-in firewall). If the device connects to the internet all captures will be lost because they will be encrypted with a key that is in the cloud and I have no access to it. Don't block the phone or the laptop, they need access to the internet.

Do this and capture the commands again (yes, sorry about that).

**Comment by Virtual-Nomad** [comment]
@felipediel No problem although will be later tonight

**Comment by Virtual-Nomad** [comment]
@felipediel Ok so a new set of results for you
Setup 
Iphone ios12 with broadlink app with RM4 Pro added as a device 
- Connected to WIfi with Internet access
RM4 Pro Connected to WfFi 
- Blocked from internet access 
- (unsure how to be 100% positive but as a test added my laptop and iphone to the same rule and they were blocked so assume it would be)

Did a number of Wireshark Captures for
- Sending Off command x 3
- Learning wIth the LongPress phase of the add function
- Learning with the LongPress + short press phases of the add function
- Learning with the LongPress + short press and Send phases of the add function

[RM4ProInternetBlock-RFOfficeFan.zip](https://github.com/mjg59/python-broadlink/files/4598015/RM4ProInternetBlock-RFOfficeFan.zip)

**Comment by felipediel** [comment]
@Virtual-Nomad It didn't work, your device is still connected to the cloud. I'll teach you how to check if it worked.

1. Open the captures in Wireshark.
2. In filter, type `ip.addr == 192.168.1.172`.
3. Find a packet that has `07` in the sixth row of the first column (beside `0050`).
4. Check the last byte of that packet. It needs to be `00`.
5. If the last byte is `01`, it didn’t work. You need to stop everything and start over.

**Comment by felipediel** [comment]
Try blocking your phone with the firewall.

**Comment by Virtual-Nomad** [comment]
I'll get another router that allows me to view packets/traffic and has better capabilities than my current home internet router. 
Any chance you can send me a screenshot of what you mean in the above? I am not sure if my columns are possibly different. Ie my 6th column is Length and I can't seem to find any 0050 (See screenshot)
![image](https://user-images.githubusercontent.com/57859302/81447659-66a41880-91c0-11ea-8735-3a265f5af06b.png)

For the heck of it I have taken another record and send with the phone blocked as well. So you never know that may work the way you need. 

[RM4Pro+PhoneInternetBlock.zip](https://github.com/mjg59/python-broadlink/files/4601551/RM4Pro%2BPhoneInternetBlock.zip)

Thanks for the patients.

**Comment by felipediel** [comment]
![wireshark](https://user-images.githubusercontent.com/41558831/81449608-01680d00-9157-11ea-87fb-755aaff203b3.png)

**Comment by Virtual-Nomad** [comment]
So I am a little confused. I have just turned my internet router off completely and done a capture, however as per the screenshot it still show 01 at the end of the packet
![image](https://user-images.githubusercontent.com/57859302/81453940-3f544800-91ce-11ea-96a8-fc45091df101.png)

I have attached the capture anyway.
[RM4Pro+PhoneInternetBlock-RFBedroomFan-LearnOff-LongPress+ShortPress+Send2.zip](https://github.com/mjg59/python-broadlink/files/4601993/RM4Pro%2BPhoneInternetBlock-RFBedroomFan-LearnOff-LongPress%2BShortPress%2BSend2.zip)

Any thoughts

**Comment by Virtual-Nomad** [comment]
SO not sure where I am going wrong. I disabeld the internet router all together but still get the following 

![image](https://user-images.githubusercontent.com/57859302/81454641-2e0c3b00-91d0-11ea-9ecb-685a724a9e7b.png)
[RM4Pro+PhoneInternetBlock-RFBedroomFan-LearnOff-LongPress+ShortPress2.zip](https://github.com/mjg59/python-broadlink/files/4602045/RM4Pro%2BPhoneInternetBlock-RFBedroomFan-LearnOff-LongPress%2BShortPress2.zip)

I have done another capture in that state just the same

**Comment by felipediel** [comment]
Hey @Virtual-Nomad, good news for you.
My code is working fine for the RF functions, I don't need to torture you anymore.

After [this update](https://github.com/mjg59/python-broadlink/pull/365) you will be able to learn RF from your RM4 Pro using the CLI. Soon I will take this changes to Home Assistant, but we still have other updates in the queue, so don't hold your breath.

**Comment by felipediel** [comment]
What worries me is the temperature and humidity. Is it showing zero? Are you using the HTS2 cable?

**Comment by Virtual-Nomad** [comment]
@felipediel sounds good ... How do I keep tabs on when it has been included in a beat or release version?

**Comment by Virtual-Nomad** [comment]
> What worries me is the temperature and humidity. Is it showing zero? Are you using the HTS2 cable?

Ok so that would be the issue then. I don't and assumed it was part of the RM4 Pro. Thanks

**Comment by felipediel** [comment]
> @felipediel sounds good ... How do I keep tabs on when it has been included in a beat or release version?

I will keep you informed here.

**Comment by Lecturer555** [comment]
Felipediel, Virtual-Nomad Hello!
 I am also really looking forward to solving this problem with RF. 
Thank you for not leaving this unattended! Waiting for an update at Home Assistents

**Comment by com-xuonghuynh** [comment]
Hope this will work soon!

**Comment by Virtual-Nomad** [comment]
@felipediel Any chance there may be an update on when we may see this built into a beta or full release I would love to test it as I need to make a call if I continue with the broadlink or move to something else. Thanks in advance.

**Comment by felipediel** [comment]
We are in the middle of a refactor with 4 updates in the queue. New features are blocked.
I also need this. Once everything is ready, I will add it, but as I said, it will take a while.

**Comment by Virtual-Nomad** [comment]
@felipediel Thanks for the response ... without much experience in code dev is it likely to be weeks or months? I certainly understand you will have priorities so I am not wishing to pressure you.

Thanks as always your skills and efforts are appreciated

**Comment by com-xuonghuynh** [comment]
@felipediel Sorry, can it work with broadlink_cli yet ? I just try to pull the new code then try to catch from broadlink_cli with my RF remote but it not catch anything.

**Comment by felipediel** [comment]
@Virtual-Nomad Sorry, I can't give you a deadline because there are steps that don't depend on me. But I promise to dedicate all my free time to these updates.

@com-xuonghuynh Are you sure? Did you use `--rfscanlearn`?

**Comment by com-xuonghuynh** [comment]
@felipediel I just did it. Press and hold the remote button until the orange light gone but "RF Frequency not found". I can add it by using Broadlink App before.

**Comment by perfektor2007** [comment]
@com-xuonghuynh out of curiosity, what is your device type and what firmware are you running (from the broadlink app)?

I tried to figure rfscanlearn for days without success and finally gave up and tried adding the RM4 pro back to the broadlink app. Tried learning RF within the app but also failed there☹️.

At the end I stumbled upon few posts from the community within the app that rf learning is broken in the last updates...
Maybe you can check it out as well.

My device type is 0x61a2 and broadlink app states firmware version v52076.

Cheers

**Comment by Virtual-Nomad** [comment]
@felipediel just checking in to see if it has now been incorporated. Or is there a way I can pull down your fix separately?

**Comment by felipediel** [comment]
@Virtual-Nomad This will take time. It is not as simple as it looks. This feature requires modifying the database structure. We are in the middle of a refactoring. I can't touch this right now.

You can use the CLI for now.

**Comment by ryanerwin** [comment]
> You can use the CLI for now

@felipediel How do use the CLI for now to learn RF signals with the RM4 pro?
Do I need to be running the latest git checkout or is the 0.14.0 latest pip3 release sufficient?

I've tried `broadlink_cli --learn` and I get: `Learning... No data received...`
I've tried `broadlink_cli --rfscanlearn` and I get: `RF Frequency not found`

Here are background details on my systems
<summary>broadlink_discovery
<details>

```
Discovering...
###########################################
RM4
# broadlink_cli --type 0x6026 --host 192.168.1.51 --mac 6c79e4a7df24
Device file data (to be used with --device @filename in broadlink_cli) : 
0x6026 192.168.1.51 6c79e4a7df24
temperature = 30.38
```
</details>
</summary>

<summary>pip3 show broadlink
<details>

```
Version: 0.14.0
Summary: Python API for controlling Broadlink IR controllers
Home-page: http://github.com/mjg59/python-broadlink
Author: Matthew Garrett
Author-email: mjg59@srcf.ucam.org
License: UNKNOWN
Location: /usr/local/lib/python3.8/dist-packages
Requires: cryptography
Required-by: 
```
</details>
</summary>

<summary>neofetch
<details>
<pre> <font color="#3971ED">❯</font> <font color="#198844">neofetch</font>
<font color="#CC342B"><b>            .-/+oossssoo+/-.</b></font>               <font color="#CC342B"><b>ryan</b></font>@<font color="#CC342B"><b>ri9</b></font> 
<font color="#CC342B"><b>        `:+ssssssssssssssssss+:`</b></font>           -------- 
<font color="#CC342B"><b>      -+ssssssssssssssssssyyssss+-</b></font>         <font color="#CC342B"><b>OS</b></font>: Ubuntu 20.04 LTS x86_64 
<font color="#CC342B"><b>    .ossssssssssssssssss</b></font><b>dMMMNy</b><font color="#CC342B"><b>sssso.</b></font>       <font color="#CC342B"><b>Kernel</b></font>: 5.4.0-37-generic 
<font color="#CC342B"><b>   /sssssssssss</b></font><b>hdmmNNmmyNMMMMh</b><font color="#CC342B"><b>ssssss/</b></font>      <font color="#CC342B"><b>Uptime</b></font>: 1 hour, 50 mins 
<font color="#CC342B"><b>  +sssssssss</b></font><b>hm</b><font color="#CC342B"><b>yd</b></font><b>MMMMMMMNddddy</b><font color="#CC342B"><b>ssssssss+</b></font>     <font color="#CC342B"><b>Packages</b></font>: 4642 (dpkg), 7 (snap) 
<font color="#CC342B"><b> /ssssssss</b></font><b>hNMMM</b><font color="#CC342B"><b>yh</b></font><b>hyyyyhmNMMMNh</b><font color="#CC342B"><b>ssssssss/</b></font>    <font color="#CC342B"><b>Shell</b></font>: zsh 5.8 
<font color="#CC342B"><b>.ssssssss</b></font><b>dMMMNh</b><font color="#CC342B"><b>ssssssssss</b></font><b>hNMMMd</b><font color="#CC342B"><b>ssssssss.</b></font>   <font color="#CC342B"><b>Resolution</b></font>: 5120x1440 
<font color="#CC342B"><b>+ssss</b></font><b>hhhyNMMNy</b><font color="#CC342B"><b>ssssssssssss</b></font><b>yNMMMy</b><font color="#CC342B"><b>sssssss+</b></font>   <font color="#CC342B"><b>DE</b></font>: GNOME 
<font color="#CC342B"><b>oss</b></font><b>yNMMMNyMMh</b><font color="#CC342B"><b>ssssssssssssss</b></font><b>hmmmh</b><font color="#CC342B"><b>ssssssso</b></font>   <font color="#CC342B"><b>WM</b></font>: Mutter 
<font color="#CC342B"><b>oss</b></font><b>yNMMMNyMMh</b><font color="#CC342B"><b>sssssssssssssshmmmhssssssso</b></font>   <font color="#CC342B"><b>WM Theme</b></font>: Adwaita 
<font color="#CC342B"><b>+ssss</b></font><b>hhhyNMMNy</b><font color="#CC342B"><b>ssssssssssss</b></font><b>yNMMMy</b><font color="#CC342B"><b>sssssss+</b></font>   <font color="#CC342B"><b>Theme</b></font>: Yaru-light [GTK2/3] 
<font color="#CC342B"><b>.ssssssss</b></font><b>dMMMNh</b><font color="#CC342B"><b>ssssssssss</b></font><b>hNMMMd</b><font color="#CC342B"><b>ssssssss.</b></font>   <font color="#CC342B"><b>Icons</b></font>: Papirus-Dark [GTK2/3] 
<font color="#CC342B"><b> /ssssssss</b></font><b>hNMMM</b><font color="#CC342B"><b>yh</b></font><b>hyyyyhdNMMMNh</b><font color="#CC342B"><b>ssssssss/</b></font>    <font color="#CC342B"><b>Terminal</b></font>: gnome-terminal 
<font color="#CC342B"><b>  +sssssssss</b></font><b>dm</b><font color="#CC342B"><b>yd</b></font><b>MMMMMMMMddddy</b><font color="#CC342B"><b>ssssssss+</b></font>     <font color="#CC342B"><b>CPU</b></font>: Intel i9-9900K (16) @ 5.000GHz 
<font color="#CC342B"><b>   /sssssssssss</b></font><b>hdmNNNNmyNMMMMh</b><font color="#CC342B"><b>ssssss/</b></font>      <font color="#CC342B"><b>GPU</b></font>: AMD ATI Radeon RX 470/480/570/5 
<font color="#CC342B"><b>    .ossssssssssssssssss</b></font><b>dMMMNy</b><font color="#CC342B"><b>sssso.</b></font>       <font color="#CC342B"><b>Memory</b></font>: 5771MiB / 31842MiB 
<font color="#CC342B"><b>      -+sssssssssssssssss</b></font><b>yyy</b><font color="#CC342B"><b>ssss+-</b></font>
<font color="#CC342B"><b>        `:+ssssssssssssssssss+:`</b></font>           <span style="background-color:#1D1F21"><font color="#1D1F21">   </font></span><span style="background-color:#CC342B"><font color="#CC342B">   </font></span><span style="background-color:#198844"><font color="#198844">   </font></span><span style="background-color:#FBA922"><font color="#FBA922">   </font></span><span style="background-color:#3971ED"><font color="#3971ED">   </font></span><span style="background-color:#A36AC7"><font color="#A36AC7">   </font></span><span style="background-color:#3971ED"><font color="#3971ED">   </font></span><span style="background-color:#C5C8C6"><font color="#C5C8C6">   </font></span>
<font color="#CC342B"><b>            .-/+oossssoo+/-.</b></font>
</pre>
</details>
</summary>

**Comment by felipediel** [comment]
@ryanerwin I think there is a problem with the latest firmware on these devices. Some users are complaining that the RF doesn't work even with the official app. If confirmed, you have wait for the next firmware update.

**Comment by ryanerwin** [comment]
> Some users are complaining that the RF doesn't work even with the official app

I had been avoiding the official app until now, but was concerned that perhaps my devices were not compatible all all... So I connected the official (Android app) to my RM4 pro and I'm able to learn most of the RF signals that I'm concerned about via the official app.

I can now confirm, with the latest firmware (`v52076`) and the [official Android app](https://play.google.com/store/apps/details?id=cn.com.broadlink.econtrol.international) (`v1.3.31`), the RF function does work on the RM4 Pro.

**Comment by Virtual-Nomad** [comment]
@felipediel ... that annoying Australian is back just checking up if there has been any movement on this? I will start fitting out the new house in 2 weeks, so probably at the point at which I need to make a call if I will wait for the inclusion of your fix, or changing platforms. What are your thoughts? Should i move to a different HA rf bridge solution?

**Comment by ryanerwin** [comment]
@Virtual-Nomad Same question....

> What are your thoughts? Should i move to a different HA RF bridge solution?

Sure wish Broadlink would actually help support @felipediel ... I messaged them to ask what sort of developer support they offer, but they want to hear from the actual project owner, not some random user

**Comment by felipediel** [comment]
@Virtual-Nomad We are in the middle of a [refactor](https://github.com/home-assistant/core/pull/36914). Then we have two updates in the queue. And then I can start working on this feature. It won't be ready in two weeks.

You can use `--rfscanlearn` as a temporary solution to obtain RF codes:

```bash
$ ./broadlink_cli --rfscanlearn --device "0x6026 192.168.1.172 24dfa7b9c05a"
Learning RF Frequency, press and hold the button to learn...
```

Press and hold a button on the remote.

```txt
Found RF Frequency - 1 of 2!
You can now let go of the button
Press enter to continue...
```

Press enter.

```txt
To complete learning, single press the button you want to learn
```

Short press the button and you get the code:

```txt
Found RF Frequency - 2 of 2!
b2002c0221000017b1121101121112112211022101221121211011222110110000000
Base64: b'sgAsAhEREhESEQIEhEBEhESERISERIRARIiEQESERISERESEhEAAF3AAAAAAAAA=='
```

Now you can set up a switch with Home Assistant:
```yaml
switch:
  - platform: broadlink
    host: 192.168.1.172
    mac: 24:df:a7:b9:c0:5a
    switches:
        name_of_the_switch:
            friendly_name: "LG Tv Power"
            command_on: sgAsAhEREhESEQIEhEBEhESERISERIRARIiEQESERISERESEhEAAF3AAAAAAAAA==
            command_off: sgAsAhEREhESEQIEhEBEhESERISERIRARIiEQESERISERESEhEAAF3AAAAAAAAA==
```

You can also use broadlink.send:
```yaml
script:
  name_of_the_script:
    sequence:
      - service: broadlink.send
        data:
          host: 192.168.1.172
          packet: sgAsAhEREhESEQIEhEBEhESERISERIRARIiEQESERISERESEhEAAF3AAAAAAAAA==

**Comment by Virtual-Nomad** [comment]
@felipediel thanks for the above .... works a treat so I can go ahead as planned. Really appreciate all the help

**Comment by burakcankose** [comment]
Hi !

I tried this but discovery OK! but learn have error.

Traceback (most recent call last):
  File "./broadlink_cli", line 8, in <module>
    import broadlink
ModuleNotFoundError: No module named 'broadlink'

**Comment by felipediel** [comment]
@burakcankose Try this:
```bash
$ python3 broadlink_cli --rfscanlearn --device "0x6026 192.168.1.172 24dfa7b9c05a"
```
If it does not work you need to install the module:
```bash
$ pip3 install broadlink
```

**Comment by burakcankose** [comment]
pip install broadlink
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: broadlink in /usr/local/lib/python2.7/dist-packages (0.14.0)
Requirement already satisfied: cryptography>=2.1.1 in /usr/lib/python2.7/dist-packages (from broadlink) (2.6.1)

**Comment by felipediel** [comment]
pip3

**Comment by Seeker100** [comment]
Hi,
First time to this forum.

I have the RM4 Pro and Veraplus hub and have the same issue.
I have downloaded on the 9th July the broadlink_cli app and the broadlink module. 
So I can learn IR codes fine but nothing on the RF. The broadlink has the orange learning light come but times out with nothing picked up.

**IR**
./broadlink_cli --type 0x6026 --host redacted --mac 24dfa7d32c05 --learn
Learning…
260c1c39371c1c1c1d….......c1c1c391c1c1c1c381b1d391c1c000d05000000000000
Base64: JgBgABsdH….......BwcOBwcHBw5HBwcHDgbHTkcHAANBQAAAAAAAA==

**RF**
./broadlink_cli --type 0x6026 --host redacted --mac 24dfa7d32c05 --rfscanlearn
Learning RF Frequency, press and hold the button to learn...
RF Frequency not found

From the previous few posts it looks to work, so I wonder what I need to do the same?

**Comment by Seeker100** [comment]
Apologies, did not pick up on the link as it stated homebridge. 
I have the veraplus hub and currently run broadlink_cli on my mac, so to get a working RF learning part I need to install the homebridge on my mac?

**Comment by kiwi-cam** [comment]
@Seeker100 No, the Homebridge link is just referencing using this Python module to workaround for learning not working in the Homebridge module.

**Comment by Seeker100** [comment]
Ah okay, but not sure if I have confused something here. From my first post (see above) I have downloaded on the the 9th July the broadlink_cli and Python module from https://github.com/mjg59/python-broadlink> The RF does not work for me (IR does) and that was what I was trying to get some clarity on how to fix?

**Comment by felipediel** [comment]
@Seeker100 Did you hold the button?

**Comment by Seeker100** [comment]
if you meant on the remote, yes I hold it down till the timeout (about 30 secs). I have just tried with two different RF remotes (one of them being the amazon firestick remote and the other a roller blind remote). The IR learning is still working.

**Comment by felipediel** [comment]
You can only control 315 and 433MHz devices. Amazon Firestick is 2.4GHz.

**Comment by Seeker100** [comment]
Ah, okay but the main roller blind remote is 433Mhz and I have managed to learn the RF through the broadlink app (I've been told by broadlink that the e-control and IHC app do not work with the RM4?)

**Comment by felipediel** [comment]
I have no idea, but it should work, since it is working for other users with the same device. Have you tried with another remote?

**Comment by Seeker100** [comment]
Unfortunately most of my other remotes use 2.4Ghz. I will try to get hold of one to test

**Comment by Seeker100** [comment]
okay, managed to find my kitchen extractor fan remote which operates at 433.92Mhz
Does not work either.

**Comment by Taylor-Carnell** [comment]
@Seeker100 
> Ah, okay but the main roller blind remote is 433Mhz and I have managed to learn the RF through the broadlink app (I've been told by broadlink that the e-control and IHC app do not work with the RM4?)

Have you got the RM4 integrated with the broadlink cloud through the app? If so, you need to remove it as it interferes with RM4 control via HASS (although I'm not sure if that applies to the python-broadlink package).
https://www.home-assistant.io/integrations/broadlink/#troubleshooting

Also, are you seeing the red light come on for the RM4? If so, does it stay on the full 30 seconds until timeout?

**Comment by Seeker100** [comment]
No the RM4 is not integrated through the cloud with any of the broadlink apps.

Yes, the light definitely stays on the full 30 secs before timing out.

**Comment by joramos21** [comment]
Hi,
Finally there is useful information. Thank you guys. I just received an RM4 Pro and it was a pain connect it to the network. I had to use the Broadlink app named "Broadlink -Universal TV Remote". Then I was able to add my ir and rf devices and use my phone to manage the devices. It works.
Something I really dislike is that it doesn't work without internet access and since yesterday, the app transfered around 12 Mb !!! This may be due to custom pictures I set for the devices.
Anyways, I don't want the device to connect to internet. I hope it can work with Home Assistant (I have HassOS 4.12, app version 2.3.0-306-full).
Someone here posted this link (thank you) : https://www.home-assistant.io/integrations/broadlink/#troubleshooting
It is written :
"Many users are experiencing problems with the new Broadlink RM Mini 3 and the entire RM4 series.
Once connected to the cloud, these devices lose their local capabilities and can no longer be controlled by Home Assistant.
To fix the problem, you need to follow these steps:
Remove the device from Broadlink App
Factory reset the device
Add the device to your local network using the app
Do not set up a cloud (not now, not ever). This means that you don’t have to complete the setup in the app, configure only the Wi-Fi and don’t add the Broadlink device to the app"

If I want to connect the Broadlink to my router, the app asks to create an account so, since it is created, it is connected to the cloud.
So I don't know how to do it.
Does someone know if this model will work on my local network only without internet accces ?

Thank you very much for your help.

Cheers.

Jo

**Comment by joramos21** [comment]
One more thing. I download on github a software that runs on Windows called "Broadlink Manager 1.0.0.1".
I start a scan and it finds the Broadlink on the network but shows as "Unknow" and when I click on "Connect", I get this :
ooking for devices...
Found device : Unknown [192.168.1.27]
Writing compatible device not detected!

**Comment by felipediel** [comment]
If you wait for the next version of Home Assistant things will be easier for you, as there will be a configuration flow to assist with unlocking the device.

I need to check your device type to make sure things will work for you. Please install this packet (`pip install broadlink`), open Python 3 and:
```python3
import broadlink as blk
devs = blk.discover(timeout=5)
[(hex(d.devtype), d.model) for d in devs]
```
What is the output?

**Comment by joramos21** [comment]
Hi,
Thank you for the answer :-)

I get : 
[]

Nothing else.

Thank you.

Good week-end.

Jo

**Comment by joramos21** [comment]
Sorry, one more thing.

If I add in the configuration.yaml file the broadlink device as a remote, I doesn't appear in the list of services.
I have to declare it as a switch. I don't know why, if there is already a switch in the file, it is not allowed and gives an error.

So I remove the other switch and added this :
switch:
  - platform: broadlink
    #friendly_name: RM4 Pro
    host: 192.168.1.27
    type: rm4_pro
    mac: 24:DF:A7:F0:E1:76
    timeout: 15

Then, I can see in the list of services on the developper tools a service called broadlink.send (in all the tutos I have found the service contains the ip address, I don't know if it means something).

When I click on the "call service" button, it returns :
Failed to call service broadlink/send. required key not provided @ data['packet'] 

Any idea ?

Thank you.

**Comment by felipediel** [comment]
I made a flow to assist with the configuration of theses devices, as it sometimes can be a little tricky. The official release will be available soon. Don't fry your brains out to make a setup that will undergo breaking changes soon. You just need to wait and things will be much simpler.

**Comment by felipediel** [comment]
I think that your device is locked btw.
```python3
import broadlink as blk
d = blk.discover(timeout=5, discover_ip_address='192.168.1.27')[0]
(hex(d.devtype), d.model, d.is_locked)
```
What is the output?

**Comment by joramos21** [comment]
('0x61a2', 'RM4 pro', False)

**Comment by joramos21** [comment]
Hi made some more tests. I added this to configuration.yaml 👍 
switch:
  - platform: broadlink
    #friendly_name: RM4 Pro
    host: 192.168.1.27
    type: rm4_pro
    mac: '24:DF:A7:F0:E1:76'
    timeout: 15

remote:
  - platform: broadlink
    #friendly_name: RM4 Pro
    host: 192.168.69.27
    type: rm4_pro
    mac: '24:DF:A7:F0:E1:76'
    timeout: 15

Then under Services (Developper tools), I type :
Service = remote.learn_command
Entity = remote.broadlink

Service Data (YAML, optional)
1 entity_id: remote.broadlink
2 device: ampli_Denon
3 command: mute

Then I clicked on button "CALL SERVICE" and I received a notification and the led on the broadlink device turned on until I pressed the mute button on my Denon amplifier.
I read that then there must be a new file in the config folder, that was not created and on the states (Developper tools), column "Entity" there must be an entry like "persistent_notification_notification" that is not present.
That's all.
Thank you

**Comment by felipediel** [comment]
> ('0x61a2', 'RM4 pro', False)

Your device is fully supported and is unlocked. About this error:

> When I click on the "call service" button, it returns :
> Failed to call service broadlink/send. required key not provided @ data['packet']

You need to specify a packet when calling `broadlink.send`:

```yaml
# Example configuration.yaml

script:
  tv_select_source:
    sequence:
      - service: broadlink.send
        data:
          host: 192.168.1.27
          packet: "JgCMAJSSFDYUNhQ2FBEUERQRFBEUERQ2FDYUNhQRFBEUERQRFBEUERQRFDYUERQRFBEUERQRFDYUNhQRFDYUNhQ2FDYUNhQABfWUkhQ2FDYUNhQRFBEUERQRFBEUNhQ2FDYUERQRFBEUERQRFBEUERQ2FBEUERQRFBEUERQ2FDYUERQ2FDYUNhQ2FDYUAA0FAAAAAAAAAAAAAAAA"
```

About this issue:
> I read that then there must be a new file in the config folder, that was not created and on the states (Developper tools), column "Entity" there must be an entry like "persistent_notification_notification" that is not present.

`remote.learn_command` does not create persistent notifications. The codes are stored in the `.storage` folder.

**Comment by joramos21** [comment]
Muchisimas gracias.
Thank you.

**Comment by felipediel** [comment]
@Virtual-Nomad Good news: https://github.com/home-assistant/core/pull/39671

**Comment by joramos21** [comment]
Hi,
I've been able to learn IR commands.
If I'm not wrong it has not been implemented yet for RF, right ?
Temperature sensor is not working as for many people. I haven't found a solution.
Thank you.
Good day.
Jo

**Comment by Taylor-Carnell** [comment]
> @Virtual-Nomad Good news: [home-assistant/core#39671](https://github.com/home-assistant/core/pull/39671)

This will be awesome. I've managed to get my setup working the the python tool, but it is an undertaking.

Cheers

**Comment by felipediel** [comment]
@joramos21 Are you using the HTS2 cable?

**Comment by joramos21** [comment]
Yes, I'm using the HTS2 cable.

**Comment by felipediel** [comment]
Show me your config.

**Comment by joramos21** [comment]
# BROADLINK
  - platform: broadlink
    #friendly_name: RM4 Pro
    host: 192.168.1.27
    type: rm4_pro
    mac: '24:DF:A7:F0:E1:76'
#    timeout: 15
    switches:
      television:
        friendly_name: "television"
        command_on: 'JgDGAFYbEA0QDQ8dDxwtHA8OEA0PDg8ODw4PDg8OHhsfGxANHxsQDRANEA0fGxANEA0PDg8OEA0QDRANHg0QHBANDwAI3lccEA0PDg8cDx0sHA8ODw4PDg8ODw4PDg8OHhweHBANHhsQDRANEA4eGxANDw4QDQ8OEA0PDg8OHwsRHBANEAAI3VgbEA0QDRAcEBstGxANEA0QDRANEA0QDQ8OHxseHBANHhwQDRANDw4eHBANEA0QDRANEA0PDg8OHg0QGxAMEQANBQ=='
        command_off: 'JgDGAFYbEA0QDQ8dDxwtHA8OEA0PDg8ODw4PDg8OHhsfGxANHxsQDRANEA0fGxANEA0PDg8OEA0QDRANHg0QHBANDwAI3lccEA0PDg8cDx0sHA8ODw4PDg8ODw4PDg8OHhweHBANHhsQDRANEA4eGxANDw4QDQ8OEA0PDg8OHwsRHBANEAAI3VgbEA0QDRAcEBstGxANEA0QDRANEA0QDQ8OHxseHBANHhwQDRANDw4eHBANEA0QDRANEA0PDg8OHg0QGxAMEQANBQ=='
#        command_mute: 'JgC0ApGUEzcSOBI3ExISExEUEhISExI4EjgSNxMSEhMTEhITEhMSNxM3EzYTNxMSExITEhITEhMSEhMSExITNxM2EzcTNxMABf+TlBI3EjcTOBISExITEhISExITNxM3EjcTEhMSExITEhITEjcTNxM2EzcTEhMSExISExITEhITEhMSEzcSNxM3EzcSAAYAk5MSOBE4EjgSExEUERQRExITEjgSOBE4EhMSExITERQRFBE4EjgSNxI4ExISExITERQRExITEhMSExI4ETgSOBI4EQAGAZKUEjgROBI4EhMSExEUERMSExI4EjgROBMSEhMSExITERQROBI4EjcSOBITEhMSExEUERMSExITEhMSOBE4EjgSOBEABgKQlRI4ETgSOBITERQQFRAUERQSOBE4EjgRFBEUEBUQFRAVEDkRORcyETkRFBEUEBUQFRAUERQRFBEUETkQORE5ETkRAAYBkZURORA5ETkRFBAVEBUQFBEUETkQORE5ERQRFBEUEBUQFRA5ETkQORE5ERQRFBAVEBUQFRAUERQRFBE5EDkRORE5EQAGAZKUEjgROBI4ExISExEUERMSExI4EjgROBITEhMSExITERQROBI4EjcSOBITEhMSExEUERMSExITEhMSOBE4EjgSOBEABgGSlBI4ETgSOBITERQRFBETERQRORA6ETgRFBEUEBUQFRAVEDkRORA5ETkRFBEUEBUQFRAUERQRFBEUETkQORE5ETkQAAYCkZURORA5ETkRFBAVEBUQFBEUETkQORE5ERQRFBAVEBUQFRA5ETkQORE5ERQRFBAVEBUQFBEUERQRFBE5EDkRORE5EAAKiFUeDg8NEA0fDR4qHw0QDRANEA0QDRANEA0QGx4cHg0QHB4cEA0QDR4NEBwPDR8bEA0fDRANEA0QGxANAA0FAAA='
#        command_volup: 'JgBAA5KTEzYTNxM2ExITEhMSExITEhM2EzcTNxMSExISEhMSExITNxM2EzcTEhMSExITEhISExITEhMSEzcSNxM3EzcSNxMABf+UkhM3EzcTNhMSExITEhMSExITNhM3EzcTEhITEhITEhMSEzcTNxI3ExITEhMSEhMSEhMSExITEhM3EjcTNxM2EzcTAAX/lJMSNxM3EzYTEhMSExITEhMSEzYTNxM3ExISExITEhITEhM3EjcTNxMSExISExEUEhITEhITExITNxI3EzcTNhM3EwAF/5OTEjgSOBI3EhMSExITEhMSExE4EjgSOBEUERQRExITEhMSOBE4EjgSExITERQRFBETEhMSExITEjcSOBI4EjcSOBIABgCSlBI4EjgSNxITEhMSExITEhMROBI4EjgRFBETEhMSExITEjgROBI4EhMSExEUERQRExITERQSExI4ETgSOBI3EjgSAAYAkpQSOBI4EzcRExITEhMSExITEDkSOBI4ERQRFBETEhMSExI4EDkSOBITEBUQFRAUEhMRFBEUERQQORI4EjgROBI4EgAGAJKVETgSOBE4ExMRExITEhMSExE4EjgSOBEUERMSExITEhMSOBE4EjgSExITERQRFBETEhMSExITEjcSOBI4ETgSOBIABgGSkxI4EjgSNxITEhMSExITEhMROBI4EjgRFBEUERMSExITEjgSNxI4EhMSExEUERMSExITEhMSExI3EjgSOBI3EjgSAAYBkpMSOBI4EjcTEhITEhMSExITETgSOBM3ExISExETEhMSExI4ETgSOBMSEhMRFBETEhMSExITEhMSNxM4ETgSNxI5EQAGAZKTEzcTNxM2ExITEhMSExITEhI3EzcTNxITEhITExISExITNxI3EzcTEhMSEhMSEhMSExITEhMSEzYTNxM3EzYTNxMABgCTkhM3EzcTNhMSExITEhMSExISNxM3EzcSExISExITEhMSEzcSNxM3ExITEhITEhITEhMSExITEhM2EzcTNxM2EzcTAAqFWBwPDg8ODxwQHCwcEA0QDRANEA0QDQ8ODw4eHB4cDw4eHB0OEA0PHQ8OHQ4PHB4ODxwQDRANEA0eDg8ADQUAAAAAAAA=''
#        command_volDOWN: 'JgC0ApOSEzcTNxI3ExITEhMSExITEhI3EzcTNhMSExITEhMSExITNhM3ExITNxMSEhITEhMSExITEhM2ExITNxM3EzYTNxMABf+UkhM3EzcSNxMSExITEhMSExISNxM3EzYTEhMSExITEhMSEzYTNxMSEzcTERMSExITEhMSExITNhMSEzcTNhQ2EzcTAAYAk5ITNxM3EzcSEhMSExITEhMSEjcTNxM2ExITEhMSExITEhM2EzcTEhM3EhITEhMSExITEhMSEzYTEhM3EzcSNxM3EwAF/5STEjcTNhM3ExITEhMSExISExI3EzcTNhMSExITEhMSExITNhM3ExITNxISExITEhMSExITEhI3ExITNxM3EjcTNxMABgCTkhM3EzYTNxMSExITEhITEhMSNxM3EzYTEhMSExITEhMSEjcTNxMSEzYTEhMSExITEhMSExISNxQSFzITNxI3EzcTAAYAk5ITNxM2EzcTEhMSExISExITEjcTNxM2ExITEhMSExITEhI3EzcTEhM3EhITEhMSExITEhITEjcTEhM3EzcSNxM3EwAF/5OTEzcTNxI3ExITEhMSEhMSExI3EzcTNhMSExITEhMSEhMSNxM3ExITNxISExITEhMSExITEhI3ExITNxM2EzcTNxMABf+TkxM3EzYTNxMSExITEhITEhMSNxM3EzYTEhQRExITEhMSEjcTNxMSEzcSEhMSExITEhMSEhMSNxMSEzcTNhM3EzcSAAYBkpMTNxM2EzcTEhMSExITEhITEjcTNxI3ExITEhMSExITEhI3EzcTEhM2ExITEhMSExITEhMSEjcTEhM3EzcSNxM3EwAKhlYdDw4PDg8cDx0sHA8ODw4PDg8ODw4PDg8OHR0dHBAOHRweDg8ODxwQDR4ODxweDg8cDw4PDg8OHg0QAA0FAAA='


# REMOTES ----------------------------------
remote:
  - platform: broadlink
    #friendly_name: RM4 Pro
    host: 192.168.1.27
    type: rm4_pro
    mac: '24:DF:A7:F0:E1:76'
    timeout: 15


# SENSORS -----------------------------------
sensor:
  - platform: broadlink
    host: 192.168.1.27
    mac: '24:DF:A7:F0:E1:76'
    monitored_conditions:
      - 'temperature'

**Comment by joramos21** [comment]
Sorry for the format. Some character seem to be interpreted.

**Comment by felipediel** [comment]
You forgot to enter the type for the sensor.

**Comment by joramos21** [comment]
Hi,
It works perfectly.
Thanks a lot.
I saw many people posting the configuration without the type so I did it wrong.
About the learning RF signals, it is not possible yet right ? It is what I understood reading your comments.
Good day to you and thank you for your availability.

**Comment by joramos21** [comment]
Hi,
Still not possible to learn a RF signal. IR works great. When I click on the "CALL SERVICE" button, the red light on the Broadlink turns on but even pressing a long time on the RF remote button, nothing happens.
Is there a way to learn RF signals?
Thank you.
Good weekend.
Jo

**Comment by felipediel** [comment]
https://www.home-assistant.io/integrations/broadlink/

**Comment by joramos21** [comment]
Hi,
I have been following both Requests and if I understand we need to wait a little bit more to have the broadlink be able to learn RF signals.
I was able to learn ir codes and suddenly (I don't know if it is due to an update) the device stopped working. I can see it on the network. I didn't change the configuration.yaml file and the ip I declared and the mac are correct. I can ping the device.
But on the "Developer Tools" > "States", when I look for the "remote.broadlink" Entity I find it but the icon is greyed out and when I select it, State = unavailable.
I've been looking on the internet for 2 weeks but cannot find what the problem is.
Any idea ?
Thank you.

**Comment by KTibow** [comment]
@joramos21 HA problem, ask on HA forum. Could be YAML has been deprecated, it's a common cause.

**Comment by joramos21** [comment]
Ok, thank you. But everythink works appart Broadlink.

**Comment by KTibow** [comment]
@joramos21 I dunno what you mean. Just make a new topic there.

**Comment by felipediel** [comment]
@joramos21 We have changed many things recently. You need to read the [docs](https://www.home-assistant.io/integrations/broadlink/), add the device via Integrations page and update your configuration. If you have any issue just ping me and I come back here to help you.

**Comment by joramos21** [comment]
Hi,
As you kindly proposed, I ask for your help.
After having updated home assistant, I found the new Broadlink integration... great :-)
I'm able to learn an ir command like this :
![image](https://user-images.githubusercontent.com/11546938/97789653-8aba4280-1bc2-11eb-825e-e6711f941a22.png)

And I'm able to send the learned command like this :
![image](https://user-images.githubusercontent.com/11546938/97789713-db31a000-1bc2-11eb-9dee-1d1462eee323.png)

And it works.

I tried to create a button but it didn't work. I don't understand how to do it.
Then regarding RF signals, I read the doc. I can see I cannot proceed the same way as ir signals. I have plugs I use to turn on and off lights.

Thanks a lot.
Jo

**Comment by joramos21** [comment]
I tried like explained in the doc to add this in configuration.yaml but I get an error when checking configuration :
script:
  Denon_Radio_Favorite1:
    sequence:
      - service: remote.send_command
        data:
          entity_id: remote.broadlink_remote
          device: "DENON AVR-3313"
          command: RadioFavorite1

**Comment by Taylor-Carnell** [comment]
> I tried like explained in the doc to add this in configuration.yaml but I get an error when checking configuration :
> script:
> Denon_Radio_Favorite1:
> ...

Has that lost the indentation when you pasted it? If it looks exactly like that, you need to fix the indentation for it to be valid yaml.

**Comment by joramos21** [comment]
I split configuration.yaml like this :
group: !include groups.yaml
automation: !include automations.yaml
**script: !include scripts.yaml**
scene: !include scenes.yaml

I think indentation is correct. In configuration.yaml I added "script:" before "Denon_Radio_Favorite1:" :
![image](https://user-images.githubusercontent.com/11546938/97791402-27381100-1bd2-11eb-966d-9a09bf77733c.png)

I also tried to put it in the scripts.yaml file removing "script:"

**Comment by felipediel** [comment]
You cannot use capital letters for scripts.

**Comment by joramos21** [comment]
So great, it works, thank you.
Regarding RF, I read about all the different methods in the doc. Let's give it a try :-)

**Comment by felipediel** [comment]
I am still working on a service for learning RF commands, I will have to rebase the PR this week. For now, you can learn them using this library (as I explained before) and send using the `b64:` prefix.

**Comment by joramos21** [comment]
Hi,
How long do you think it will take to you ? Your work is great and I prefer to use your service, something reliable.
Thank you.

**Comment by felipediel** [comment]
The [PR](https://github.com/home-assistant/core/pull/39671) is ready, now we are waiting for a review.

**Comment by joramos21** [comment]
I'm new to HA so won't be the best to test but I don't understand how to learn the RF signal.
Isn't it possible to learn from Developer Tools > Services?
I have RF plugs but these don't exist within HA as devices, and no entity_id.
I tried your example but in my scripts.yaml file removing "script:" and moving code two spaces to the left to respect indentation.

**Comment by felipediel** [comment]
It will be possible after the release with the option `command_type: rf`.

**Comment by joramos21** [comment]
Hi Felipe,
Any news regarding RF implementation ?
I didn't see anything new.
Thank you very much.
Jo

**Comment by felipediel** [comment]
Hi. It is supported now. You need to specify `command_type: rf` when learning.

```yaml
script:
  learn_car_lock:
    sequence:
      - service: remote.learn_command
        data:
          entity_id: remote.garage
          device: car
          command: lock
          command_type: rf
```

**Comment by joramos21** [comment]
So great, I have been waiting for this youuuu :-)

I tried like to learn an ir code which is the easiest method for me, under development tools.

I chose remote.learn_command as Service.

Then remote.broadlink_remote as Entity

Then added under Services data :

entity_id: remote.broadlink_remote
device: kitchen_lamp
command: powerON
command_type: rf

It didn't work. I clicked on Call Service button. The red light on the device turned on, I pressed twice on my rf remote that turned the light/lamp on as expected but the red light stayed on around 15-20 seconds. 

I then tried to call the service with remote.send_command but nothing happened.

I don't know how to find the broadlink integration version that is installed. It doesn't display that an update is needed. 

Thank you. 

Jo

**Comment by felipediel** [comment]
You need to press the button at the right time. Check the notifications during the process, they will guide you.

**Comment by joramos21** [comment]
I tried again and no luck, even pressing many times, once the red led turn on, an another test waiting longer, and others tests pressing twice the button like you suggested, and more. I got the notification saying to press the button and nothing happens. With IR, when I press on the remote, then the red led turns off. Here, with RF, the led stays turned on 15-20 seconds and the code is not learned. I don't know where I could have a look at logs or a config file. Remote frequency is 433.92 Mhz. I thought maybe it is not compatible even it would suprise me. I read on amazon website, someone who bought a RM Pro+ saying it works with Intertechno Brand, which is what I have.
No error messages notifications are displayed. It's like the Broadlink was not catching the signal.

**Comment by joramos21** [comment]
One more thing. I jus tried with "command_type: rf" but pressed a button on an ir remote and it worked. I don't know if this is normal, it shouldn't. Maybe latest integration is not installed.

**Comment by felipediel** [comment]
Hi @joramos21. I just checked the master branch and realized that the update hasn't reached you yet. Sorry, we still need to wait a little, now it won't be long.

**Comment by joramos21** [comment]
No problem. Better it is that.

**Comment by joramos21** [comment]
Hi,

I've been waiting a while and have seen that someone was able to learn RF codes within HA like below but prior pressing the CALL SERVICE button the guy was using at the same time the Broadlink app to tell the Broadlink pro to set as RF.

So I did this under development tools.

I chose remote.learn_command as Service.
Then remote.broadlink_remote as Entity
Then added under Services data :

entity_id: remote.broadlink_remote
device: kitchen_lamp
command: powerON
command_type: rf

It didn't work. I clicked on Call Service button. The red light on the device turned on, I pressed on the rf remote that turned the light/lamp on (and it turned on as expected) but the red light stayed on around 15-20 seconds and the command wasn't learned. 

Is it ready now ?

Thank you.

Good day.

Jo

**Comment by joramos21** [comment]
I answer to myself. I found a solution that works for me using the Android Broadlink app (that I wanted avoid using). You can see explanations here https://community.home-assistant.io/t/guide-how-to-learn-broadlink-rf-codes/62119/152, post 150. As a summary :
1. go to Developpers to learn the RF code
2. go to the application and be prepared to learn an RF code
3. Press on the "CALL SERVICE" button then press the button on the Broadlink app to learn the code and finally, press on the remote
Cheers :-)

**Comment by felipediel** [comment]
RF features are now supported in Home Assistant. Closing.

**Comment by Kavish85** [comment]
Hi @felipediel. I have tried all whichever way and the RF simply does not work.

I open the Broadlink App on my phone, then initiate remote.learn on HA and call service. I get as far as “press & hold the button X” which I keep pressed even after the notification disappears. I don’t ever see the second “press the button X".

Is something perhaps broken or am I doing something wrong?

**Comment by felipediel** [comment]
You don't need the Broadlink app. Use remote.learn_command in Home Asssistant with the 'command_type: rf' attribute. Don't hold the button for too long. You need to follow the notifications to know when to stop and press the button again in the end.

**Comment by Kavish85** [comment]
I've got it setup as follows and then click "Call Service" and immediately follow the notification on HA by pressing the button on the Broadlink light switch and still nothing. Any other suggestions?

service: remote.learn_command
target:
  entity_id: remote.wi_fi_universal_remote_remote
  device_id: 1815da26aedbcb5d65a2e7984621c35c
data:
  device: kitchen lights
  command_type: rf
  command: 'on'

**Comment by felipediel** [comment]
Don't use 'on' for the command, this is a reserved word in YAML.

**Comment by Kavish85** [comment]
Should I leave out the quotation marks from 'on'?

**Comment by felipediel** [comment]
Don't use this word. Change it to 'turn on'.

**Comment by Kavish85** [comment]
Sadly that doesn't seem to work. Also tried 'power on', and tried turn on and power on without the inverted commas.

**Comment by Kavish85** [comment]
Silly question. Should I copy and paste this into the configuration.yaml file prior to learning the command?

service: remote.learn_command
target:
entity_id: remote.wi_fi_universal_remote_remote
device_id: 1815da26aedbcb5d65a2e7984621c35c
data:
device: kitchen lights
command_type: rf
command: 'turn on'

**Comment by Kavish85** [comment]
After 3 weeks of Broadlink's crappy process that just doesn't work, it made sense to move to Eachen smart switches. Discovered within HA and setup in 6 minutes.

**Comment by felipediel** [comment]
Awesome, another happy customer. Have a good one.
